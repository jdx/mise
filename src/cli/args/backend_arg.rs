use heck::ToKebabCase;
use std::collections::HashMap;
use std::fmt::{Debug, Display};
use std::hash::Hash;
use std::path::PathBuf;

use once_cell::sync::Lazy;

use crate::backend::{unalias_backend, BackendType};
use crate::dirs;
use crate::registry::REGISTRY;

#[derive(Clone, PartialOrd, Ord)]
pub struct BackendArg {
    /// user-specified identifier, "node", "npm:prettier", "cargo:eza", "vfox:version-fox/vfox-nodejs"
    /// multiple ids may point to a single tool, e.g.: "node", "core:node" or "vfox:version-fox/vfox-nodejs"
    /// and "vfox:https://github.com/version-fox/vfox-nodejs"
    pub id: String,
    /// the name of the tool within the backend, e.g.: "node", "prettier", "eza", "vfox-nodejs"
    pub name: String,
    /// type of backend, "asdf", "cargo", "core", "npm", "vfox"
    pub backend_type: BackendType,
    /// ~/.local/share/mise/cache/<THIS>
    pub cache_path: PathBuf,
    /// ~/.local/share/mise/installs/<THIS>
    pub installs_path: PathBuf,
    /// ~/.local/share/mise/downloads/<THIS>
    pub downloads_path: PathBuf,
}

impl<A: AsRef<str>> From<A> for BackendArg {
    fn from(s: A) -> Self {
        let s = s.as_ref();
        if let Some(fa) = FORGE_MAP.get(s) {
            return fa.clone();
        }
        if let Some((backend_type, name)) = s.split_once(':') {
            if let Ok(backend_type) = backend_type.parse() {
                return Self::new(backend_type, name);
            }
        }
        Self::new(BackendType::Asdf, s)
    }
}

impl BackendArg {
    pub fn new(backend_type: BackendType, name: &str) -> Self {
        let name = unalias_backend(name).to_string();
        let id = match backend_type {
            BackendType::Asdf | BackendType::Core => name.clone(),
            backend_type => format!("{backend_type}:{name}"),
        };
        let pathname = id.to_kebab_case();
        Self {
            name,
            backend_type,
            id,
            cache_path: dirs::CACHE.join(&pathname),
            installs_path: dirs::INSTALLS.join(&pathname),
            downloads_path: dirs::DOWNLOADS.join(&pathname),
        }
    }
}

impl Display for BackendArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.id)
    }
}

impl Debug for BackendArg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, r#"BackendArg("{}")"#, self.id)
    }
}

impl PartialEq for BackendArg {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl Eq for BackendArg {}

impl Hash for BackendArg {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.id.hash(state);
    }
}

static FORGE_MAP: Lazy<HashMap<&'static str, BackendArg>> = Lazy::new(|| {
    REGISTRY
        .iter()
        .map(|(short, full)| {
            let (backend_type, name) = full.split_once(':').unwrap();
            let backend_type = backend_type.parse().unwrap();
            let fa = BackendArg::new(backend_type, name);
            (*short, fa)
        })
        .collect()
});

#[cfg(test)]
mod tests {
    use pretty_assertions::{assert_eq, assert_str_eq};

    use super::*;

    #[test]
    fn test_backend_arg() {
        let t = |s: &str, id, name, t| {
            let fa: BackendArg = s.into();
            assert_str_eq!(fa.id, id);
            assert_str_eq!(fa.name, name);
            assert_eq!(fa.backend_type, t);
        };
        let asdf = |s, id, name| t(s, id, name, BackendType::Asdf);
        let cargo = |s, id, name| t(s, id, name, BackendType::Cargo);
        // let core = |s, id, name| t(s, id, name, BackendType::Core);
        let npm = |s, id, name| t(s, id, name, BackendType::Npm);

        asdf("asdf:poetry", "poetry", "poetry");
        asdf("poetry", "poetry", "poetry");
        asdf("", "", "");
        cargo("cargo:eza", "cargo:eza", "eza");
        // core("node", "node", "node");
        npm("npm:@antfu/ni", "npm:@antfu/ni", "@antfu/ni");
        npm("npm:prettier", "npm:prettier", "prettier");
    }

    #[test]
    fn test_backend_arg_pathname() {
        let t = |s: &str, expected| {
            let fa: BackendArg = s.into();
            let actual = fa.installs_path.to_string_lossy();
            let expected = dirs::INSTALLS.join(expected);
            assert_str_eq!(actual, expected.to_string_lossy());
        };
        t("asdf:node", "node");
        t("node", "node");
        t("", "");
        t("cargo:eza", "cargo-eza");
        t("npm:@antfu/ni", "npm-antfu-ni");
        t("npm:prettier", "npm-prettier");
    }
}
