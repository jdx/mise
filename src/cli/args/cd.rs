use clap::{Arg, ArgAction};

#[derive(Clone)]
pub struct Cd;

impl Cd {
    pub fn arg() -> Arg {
        Arg::new("cd")
            .short('C')
            .long("cd")
            .help("Change directory before running command")
            .global(true)
            .action(ArgAction::Set)
            .value_hint(clap::ValueHint::DirPath)
            .value_name("DIR")
    }
}
