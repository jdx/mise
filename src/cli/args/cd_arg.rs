use clap::{Arg, ArgAction};
use std::path::PathBuf;

#[derive(Clone)]
pub struct CdArg;

impl CdArg {
    pub fn arg() -> Arg {
        Arg::new("cd")
            .value_parser(clap::value_parser!(PathBuf))
            .short('C')
            .long("cd")
            .help("Change directory before running command")
            .global(true)
            .action(ArgAction::Set)
            .value_hint(clap::ValueHint::DirPath)
            .value_name("DIR")
    }
}
