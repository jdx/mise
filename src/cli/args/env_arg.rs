use clap::{Arg, ArgAction};
use std::sync::LazyLock as Lazy;

#[derive(Clone, Debug)]
pub struct ProfileArg;

pub static PROFILE_ARG: Lazy<Arg> = Lazy::new(ProfileArg::arg);
pub static ENV_ARG: Lazy<Arg> =
    Lazy::new(|| clap::arg!(-E --env "Set the environment for loading mise.<ENV>.toml files."));

impl ProfileArg {
    fn arg() -> Arg {
        Arg::new("profile")
            .short('P')
            .long("profile")
            .help("Set the profile (environment)")
            .value_parser(clap::value_parser!(String))
            .value_name("PROFILE")
            .action(ArgAction::Set)
            .global(true)
    }
}
