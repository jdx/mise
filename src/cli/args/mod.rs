pub use cd_arg::CdArg;
pub use env_var_arg::EnvVarArg;
pub use forge_arg::ForgeArg;
pub use log_level_arg::{DebugArg, LogLevelArg, TraceArg};
pub use quiet_arg::QuietArg;
pub use tool_arg::ToolArg;
pub use tool_arg::ToolVersionType;
pub use verbose_arg::VerboseArg;
pub use yes_arg::YesArg;

mod cd_arg;
mod env_var_arg;
mod forge_arg;
mod log_level_arg;
mod quiet_arg;
mod tool_arg;
mod verbose_arg;
mod yes_arg;
