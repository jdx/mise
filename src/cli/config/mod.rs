use clap::Subcommand;
use eyre::Result;

mod generate;
mod ls;

/// [experimental] Manage config files
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "cfg")]
pub struct Config {
    #[clap(subcommand)]
    command: Option<Commands>,

    /// Do not print table header
    #[clap(long, alias = "no-headers", verbatim_doc_comment)]
    no_header: bool,
}

#[derive(Debug, Subcommand)]
enum Commands {
    Ls(ls::ConfigLs),
    Generate(generate::ConfigGenerate),
}

impl Commands {
    pub fn run(self) -> Result<()> {
        match self {
            Self::Ls(cmd) => cmd.run(),
            Self::Generate(cmd) => cmd.run(),
        }
    }
}

impl Config {
    pub fn run(self) -> Result<()> {
        let cmd = self.command.unwrap_or(Commands::Ls(ls::ConfigLs {
            no_header: self.no_header,
        }));

        cmd.run()
    }
}
