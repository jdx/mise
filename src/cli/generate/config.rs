use crate::Result;
use crate::cli::config::generate;

/// [experimental] Generate a mise.toml file
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = generate::AFTER_LONG_HELP)]
pub struct Config {
    #[clap(flatten)]
    generate: generate::ConfigGenerate,
}

impl Config {
    pub async fn run(self) -> Result<()> {
        self.generate.run()
    }
}
