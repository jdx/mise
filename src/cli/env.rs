use eyre::Result;

use crate::cli::args::ToolArg;
use crate::config::Config;
use crate::shell::{get_shell, ShellType};
use crate::toolset::{InstallOptions, Toolset, ToolsetBuilder};

/// Exports env vars to activate mise a single time
///
/// Use this if you don't want to permanently install mise. It's not necessary to
/// use this if you have `mise activate` in your shell rc file.
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "e", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Env {
    /// Tool(s) to use
    #[clap(value_name = "TOOL@VERSION")]
    tool: Vec<ToolArg>,

    /// Output in JSON format
    #[clap(long, short = 'J', overrides_with = "shell")]
    json: bool,

    /// Shell type to generate environment variables for
    #[clap(long, short, overrides_with = "json")]
    shell: Option<ShellType>,
}

impl Env {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        let mut ts = ToolsetBuilder::new().with_args(&self.tool).build(&config)?;
        ts.install_arg_versions(&config, &InstallOptions::new())?;
        ts.notify_if_versions_missing();

        if self.json {
            self.output_json(&config, ts)
        } else {
            self.output_shell(&config, ts)
        }
    }

    fn output_json(&self, config: &Config, ts: Toolset) -> Result<()> {
        let env = ts.env_with_path(config)?;
        miseprintln!("{}", serde_json::to_string_pretty(&env)?)?;
        Ok(())
    }

    fn output_shell(&self, config: &Config, ts: Toolset) -> Result<()> {
        let default_shell = get_shell(Some(ShellType::Bash)).unwrap();
        let shell = get_shell(self.shell).unwrap_or(default_shell);
        for (k, v) in ts.env_with_path(config)? {
            let k = k.to_string();
            let v = v.to_string();
            miseprint!("{}", shell.set_env(&k, &v))?;
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  $ <bold>eval "$(mise env -s bash)"</bold>
  $ <bold>eval "$(mise env -s zsh)"</bold>
  $ <bold>mise env -s fish | source</bold>
  $ <bold>execx($(mise env -s xonsh))</bold>
"#
);

#[cfg(test)]
mod tests {
    use crate::cli::tests::grep;
    use std::env;

    use crate::dirs;

    #[test]
    fn test_env() {
        let stdout = assert_cli!("env", "-s", "bash");
        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3/bin")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_with_runtime_arg() {
        assert_cli!("install", "tiny@3.0");
        let stdout = assert_cli!("env", "tiny@3.0", "-s", "bash");

        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3.0/bin")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_alias() {
        assert_cli!("plugin", "add", "tiny");
        assert_cli!("install", "tiny@my/alias");
        let stdout = assert_cli!("env", "tiny@my/alias", "-s", "bash");
        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3.0.1")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_tiny() {
        let stdout = assert_cli!("env", "tiny@2", "tiny@1", "tiny@3", "-s", "bash");
        assert_str_eq!(grep(stdout, "JDXCODE"), "export JDXCODE_TINY=2.1.0");
    }

    #[test]
    fn test_env_default_shell() {
        env::set_var("SHELL", "");
        let stdout = assert_cli!("env");
        assert!(stdout.contains("export PATH="));
    }

    #[test]
    fn test_env_json() {
        assert_cli_snapshot!("env", "-J");
    }
}
