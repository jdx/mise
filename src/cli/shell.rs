use color_eyre::eyre::{eyre, Result};
use console::style;

use crate::cli::args::ToolArg;
use crate::config::Config;
use crate::shell::get_shell;
use crate::toolset::{InstallOptions, ToolSource, ToolsetBuilder};

/// Sets a tool version for the current session
///
/// Only works in a session where mise is already activated.
///
/// This works by setting environment variables for the current shell session
/// such as `MISE_NODE_VERSION=20` which is "eval"ed as a shell function created
/// by `mise activate`.
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, visible_alias = "sh", after_long_help = AFTER_LONG_HELP)]
pub struct Shell {
    /// Tool(s) to use
    #[clap(value_name = "TOOL@VERSION")]
    tool: Vec<ToolArg>,

    /// Number of jobs to run in parallel
    /// [default: 4]
    #[clap(long, short, env = "MISE_JOBS", verbatim_doc_comment)]
    jobs: Option<usize>,

    /// Directly pipe stdin/stdout/stderr from plugin to user
    /// Sets --jobs=1
    #[clap(long, overrides_with = "jobs")]
    raw: bool,

    /// Removes a previously set version
    #[clap(long, short)]
    unset: bool,
}

impl Shell {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        if !config.is_activated() {
            err_inactive()?;
        }

        let mut ts = ToolsetBuilder::new().with_args(&self.tool).build(&config)?;
        let opts = InstallOptions {
            force: false,
            jobs: self.jobs,
            raw: self.raw,
            latest_versions: false,
        };
        ts.install_arg_versions(&config, &opts)?;
        ts.notify_if_versions_missing();

        let shell = get_shell(None).expect("no shell detected");

        for (p, tv) in ts.list_current_installed_versions() {
            let source = &ts.versions.get(p.fa()).unwrap().source;
            if matches!(source, ToolSource::Argument) {
                let k = format!("MISE_{}_VERSION", p.id().to_uppercase());
                let op = if self.unset {
                    shell.unset_env(&k)
                } else {
                    shell.set_env(&k, &tv.version)
                };
                miseprintln!("{op}")?;
            }
        }

        Ok(())
    }
}

fn err_inactive() -> Result<()> {
    Err(eyre!(formatdoc!(
        r#"
                mise is not activated in this shell session.
                Please run `{}` first in your shell rc file.
                "#,
        style("mise activate").yellow()
    )))
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    $ <bold>mise shell node@20</bold>
    $ <bold>node -v</bold>
    v20.0.0
"#
);

#[cfg(test)]
mod tests {
    use std::env;

    #[test]
    fn test_shell() {
        let err = assert_cli_err!("shell", "tiny@1.0.1");
        assert_display_snapshot!(err);
        env::set_var("__MISE_DIFF", "");
        env::set_var("MISE_SHELL", "zsh");
        assert_cli_snapshot!("shell", "tiny@1.0.1");
        env::remove_var("__MISE_DIFF");
        env::remove_var("MISE_SHELL");
    }
}
