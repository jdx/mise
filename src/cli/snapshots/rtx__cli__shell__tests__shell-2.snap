---
source: src/cli/shell.rs
expression: output
---
export RTX_TINY_VERSION=1.0.1


